### ISSUE TITLE
Possible memory leak on Windows 10 / Server 2016 systems when using ole.CoInitializeEx and WbemScripting.SWbemLocator in a loop

### ISSUE THREAD

**Issue by gbrayut** [go-ole/go-ole#135]
I've spent a few weeks trying to track down a memory leak in our scollector monitoring agent, which uses the [github.com/StackEchange/wmi package](https://github.com/StackExchange/wmi/pull/21) based on go-ole. Our application appears to leak ~1.5MB of private working set memory for every 10,000 calls to `ole.CoInitializeEx` and `oleutil.CreateObject("WbemScripting.SWbemLocator")`, which means we leak ~500MB every few days (currently remedied by a kill switch that restarts the service after a certain limit). So far my findings are:

1. The leak only occurs on Windows 10 or Server 2016. Running the same binary on older OSes does not leak any memory.
1. A minimal repo case requires just the `ole.CoInitializeEx(...)` `oleutil.CreateObject("WbemScripting.SWbemLocator")` and `unk.QueryInterface(ole.IID_IDispatch)` calls in a loop with 50,000-250,000 iterations. On older systems the private working set will stabilize below 10-15MB, but on the newer OS it will continue to grow until the process or loop is stopped.
1. I was NOT able to reproduce the leak using `oleutil.CreateObject("Microsoft.XMLHTTP")`, `IDispatch`, and `oleutil.CallMethod(xmlhttp, "open", "GET", url, false)`. In our original testing I though it also leaked, but it was really our use of the wmi package to get the private working set size that was still leaking. After moving the memory checking function to use wmic.exe I cannot repo this leak creating Microsoft.XMLHTTP objects.
1. Our agent executes WMI queries from a large number of goroutines (10-50 usually), but I can recreate the leak using a single threaded for loop.
1. I tested using both `COINIT_APARTMENTTHREADED` and `COINIT_MULTITHREADED` modes. In both cases if the CoInitializeEx/CoUninitialize call is inside the loop it will leak memory. If it is outside the loop the memory usage is much more stable.
1.  To the best of my abilities I attempted to recreate the leak in [python using the comtypes package](https://gist.github.com/gbrayut/0271db792e297cda7bc6d5dff237654b), but I was not able to get CoInitializeEx/CreateObject/QueryInterface to cause a leak from python. We initially thought it may be an issue or change in the new OS, but have not been able to find any reports or recreate the issue outside the go-ole package.

Here is the minimal repo I have come up with in a file named memory_test.go.

```go
// Run test using: go test -run TestMemoryLeak -timeout 600m
package wmi

import (
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"testing"
	"time"

	ole "github.com/go-ole/go-ole"
	"github.com/go-ole/go-ole/oleutil"
)

func QueryWMI(query string, dst interface{}, connectServerArgs ...interface{}) error {

	//Start Section: Moving this to outside the loop prevents the memory leak in Win10/2016
	err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
	if err != nil {
		oleCode := err.(*ole.OleError).Code()
		if oleCode != ole.S_OK && oleCode != S_FALSE {
			//t.Fatal(err)
			panic(err)
		}
	}
	defer ole.CoUninitialize()
	//End Section

	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
	if err != nil {
		panic(err)
	}
	defer unknown.Release()

	wmi, err := unknown.QueryInterface(ole.IID_IDispatch)
	if err != nil {
		panic(err)
	}
	defer wmi.Release()

	serviceRaw, err := oleutil.CallMethod(wmi, "ConnectServer", connectServerArgs...)
	if err != nil {
		return err
	}
	//service := serviceRaw.ToIDispatch()
	defer serviceRaw.Clear()
	return nil
}

func TestMemoryLeak(t *testing.T) {
	defer func() {
		if r := recover(); r != nil {
			t.Error(r)
		}
	}()

	start := time.Now()
	limit := 50000000
	fmt.Printf("Benchmark Iterations: %d\n", limit)

	/*
		err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
		if err != nil {
			oleCode := err.(*ole.OleError).Code()
			if oleCode != ole.S_OK && oleCode != S_FALSE {
				//t.Fatal(err)
				panic(err)
			}
		}
		defer ole.CoUninitialize()
	*/

	for i := 0; i < limit; i++ {
		QueryWMI("", nil)
		if i%2000 == 0 {
			//Run the following to get memory usage: WMIC.exe path Win32_PerfRawData_PerfProc_Process where "IDProcess=2412" get WorkingSetPrivate /value
			out, _ := exec.Command("wmic.exe", "path", "Win32_PerfRawData_PerfProc_Process", "where", fmt.Sprintf("IDProcess=%d", os.Getpid()), "get", "WorkingSetPrivate", "/value").Output()
			str := strings.Trim(string(out), " \n\r")
			bytes, _ := strconv.Atoi(str[18:])
			fmt.Printf("Time: %4ds  Count: %7d   PrivateWS: %.2fMB\n", time.Now().Sub(start)/time.Second, i, float64(bytes)/1048576)
		}

	}
}
```

output on Windows Server 2016:
```
go test -run TestMemoryLeak -timeout 600m
Benchmark Iterations: 50000000
Time:    0s  Count:       0   PrivateWS: 2.71MB
Time:   17s  Count:    2000   PrivateWS: 5.49MB
Time:   32s  Count:    4000   PrivateWS: 8.41MB
Time:   80s  Count:    6000   PrivateWS: 10.80MB
Time:   93s  Count:    8000   PrivateWS: 13.02MB
Time:  106s  Count:   10000   PrivateWS: 14.81MB
Time:  120s  Count:   12000   PrivateWS: 16.39MB
Time:  133s  Count:   14000   PrivateWS: 17.86MB
Time:  146s  Count:   16000   PrivateWS: 19.15MB
Time:  159s  Count:   18000   PrivateWS: 20.68MB
Time:  172s  Count:   20000   PrivateWS: 21.98MB
Time:  185s  Count:   22000   PrivateWS: 23.38MB
Time:  198s  Count:   24000   PrivateWS: 25.24MB
```

output on Windows Server 2012 R2:
```
PS C:\Users\gbray\go\src\github.com\StackExchange\wmi> go test -run TestMemoryLeak -timeout 600m
Benchmark Iterations: 50000000
Time:    0s  Count:       0   PrivateWS: 1.60MB
Time:   24s  Count:    2000   PrivateWS: 3.00MB
Time:   50s  Count:    4000   PrivateWS: 4.16MB
Time:   77s  Count:    6000   PrivateWS: 5.05MB
Time:  101s  Count:    8000   PrivateWS: 5.75MB
Time:  126s  Count:   10000   PrivateWS: 6.38MB
...
Time:  258s  Count:   20000   PrivateWS: 7.45MB
...
Time:  389s  Count:   30000   PrivateWS: 7.52MB
...
Time:  518s  Count:   40000   PrivateWS: 7.54MB
...
Time:  653s  Count:   50000   PrivateWS: 7.52MB
```

We plan on refactoring the stackexchange/wmi package to use a dedicated thread for executing wmi queries so we only have to call ole.CoInitializeEx once. This should fix our issue, but I thought I would file this issue as a heads up/warning that calling ole.CoInitializeEx and WbemScripting.SWbemLocator in a loop may leak memory on Windows 10/Server 2016.

**Comment by mattn** [comment]
Could you please try to add

```
defer serviceRaw.ToIDispatch().Release()
```

**Comment by gbrayut** [comment]
```
	//service := serviceRaw.ToIDispatch()
	defer serviceRaw.Clear()
	defer serviceRaw.ToIDispatch().Release()
	return nil
```
With the above we still see a leak on Win10/2016 but no leak on 2012R2 (havent tested Win7/8/8.1):

```
go test -run TestMemoryLeak -timeout 600m
Benchmark Iterations: 50000000
Time:    0s  Count:       0   PrivateWS: 1.86MB
Time:   33s  Count:    2000   PrivateWS: 4.67MB
Time:   63s  Count:    4000   PrivateWS: 7.14MB
Time:   94s  Count:    6000   PrivateWS: 9.53MB
Time:  125s  Count:    8000   PrivateWS: 11.59MB
Time:  156s  Count:   10000   PrivateWS: 13.77MB
Time:  189s  Count:   12000   PrivateWS: 15.10MB
Time:  221s  Count:   14000   PrivateWS: 16.68MB
Time:  252s  Count:   16000   PrivateWS: 18.31MB
Time:  282s  Count:   18000   PrivateWS: 20.15MB
Time:  311s  Count:   20000   PrivateWS: 21.74MB
Time:  342s  Count:   22000   PrivateWS: 23.04MB
Time:  372s  Count:   24000   PrivateWS: 24.59MB
Time:  404s  Count:   26000   PrivateWS: 25.92MB
Time:  438s  Count:   28000   PrivateWS: 27.21MB
Time:  468s  Count:   30000   PrivateWS: 28.75MB
Time:  499s  Count:   32000   PrivateWS: 30.05MB
Time:  528s  Count:   34000   PrivateWS: 31.60MB
Time:  552s  Count:   36000   PrivateWS: 32.91MB
Time:  574s  Count:   38000   PrivateWS: 34.46MB
Time:  605s  Count:   40000   PrivateWS: 35.75MB
```

**Comment by mattn** [comment]
It seems that you put exec.Command in the code. It it same without them?

**Comment by gbrayut** [comment]
Yes... That is just to retrieve the size of the working set. Running the test without that and using task manager or resource monitor shows the same linear increase in memory usage.

**Comment by ws6** [comment]
up Windows 7 client and server 2014 SP2 found leaking too

**Comment by mattn** [comment]
please try this 

```go
package wmi

import (
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"testing"
	"time"

	ole "github.com/go-ole/go-ole"
	"github.com/go-ole/go-ole/oleutil"
)

func QueryWMI(query string, dst interface{}, connectServerArgs ...interface{}) error {

	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
	if err != nil {
		panic(err)
	}
	defer unknown.Release()

	wmi, err := unknown.QueryInterface(ole.IID_IDispatch)
	if err != nil {
		panic(err)
	}
	defer wmi.Release()

	serviceRaw, err := oleutil.CallMethod(wmi, "ConnectServer", connectServerArgs...)
	if err != nil {
		return err
	}
	//service := serviceRaw.ToIDispatch()
	defer serviceRaw.Clear()
	return nil
}

func TestMemoryLeak(t *testing.T) {
	defer func() {
		if r := recover(); r != nil {
			t.Error(r)
		}
	}()

	//Start Section: Moving this to outside the loop prevents the memory leak in Win10/2016
	err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
	if err != nil {
		oleCode := err.(*ole.OleError).Code()
		if oleCode != ole.S_OK {
			//t.Fatal(err)
			panic(err)
		}
	}
	defer ole.CoUninitialize()
	//End Section

	start := time.Now()
	limit := 50000000
	fmt.Printf("Benchmark Iterations: %d\n", limit)

	/*
		err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
		if err != nil {
			oleCode := err.(*ole.OleError).Code()
			if oleCode != ole.S_OK && oleCode != S_FALSE {
				//t.Fatal(err)
				panic(err)
			}
		}
		defer ole.CoUninitialize()
	*/

	for i := 0; i < limit; i++ {
		QueryWMI("", nil)
		if i%2000 == 0 {
			//Run the following to get memory usage: WMIC.exe path Win32_PerfRawData_PerfProc_Process where "IDProcess=2412" get WorkingSetPrivate /value
			out, _ := exec.Command("wmic.exe", "path", "Win32_PerfRawData_PerfProc_Process", "where", fmt.Sprintf("IDProcess=%d", os.Getpid()), "get", "WorkingSetPrivate", "/value").Output()
			str := strings.Trim(string(out), " \n\r")
			bytes, _ := strconv.Atoi(str[18:])
			fmt.Printf("Time: %4ds  Count: %7d   PrivateWS: %.2fMB\n", time.Now().Sub(start)/time.Second, i, float64(bytes)/1048576)
			runtime.Gosched()
		}
	}
}
```

**Comment by ws6** [comment]
Thanks @Mattn, I will try it tomorrow. 
At meanwhile, I found this, https://support.microsoft.com/en-us/help/2545479/a-memory-leak-occurs,-and-client-requests-may-take-a-long-time,-when-you-run-an-application-that-is-based-on-com-and-that-is-registered-as-single-threaded

could ole32.dll hotfix is an alter solution?

Last, do you have updates on https://github.com/mattn/go-adodb/blob/master/adodb.go as well?
sorry too many questions.
best,

**Comment by mattn** [comment]
I don't make sure the hotfix, but adding `runtime.Gosched` seems to be stop liner growing-up memories. (see above code)

**Comment by ws6** [comment]
Go1.7 Windows 7 SP1. still seeing the leaking patten.

**Comment by mattn** [comment]
Did you try I said in above?

**Comment by gbrayut** [comment]
@mattn your example does not include the call to ole.CoInitializeEx inside the loop. If that call is inside the loop we see TestMemoryLeak leaking with or without calling runtime.Gosched() inside the loop.

In our case, there are a bunch of different go routines executing WMI queries, so we have to call CoInitializeEx for each one as there is no way to launch a new go routine on an already initialized thread. We have been testing a [new design](https://github.com/bosun-monitor/bosun/pull/2028) where we moved the initialization and OLE logic to a background worker and use channels to send query requests to the background SWbemServices worker. So far that approach is working and we do NOT see any memory issues anymore when using the code from https://github.com/StackExchange/wmi/pull/23

Also, we saw one instance of the same leak on a 2012 R2 Server that was updated to [Windows Management Framework (WMF) 5.1](https://msdn.microsoft.com/en-us/powershell/wmf/5.1/release-notes), which is the version that is included in Server 2016/Windows 10. So this may not be just an OS issue, you would also need to see what $PSVersionTable is in a PowerShell prompt (5.1 and above may have the same leak).

**TL;DR:** If you frequently use ole.CoInitializeEx and WbemScripting.SWbemLocator inside of many goroutines and are using WMI 5.1 (built in to Win10/Server2016) you may experience a memory leak.

**Comment by gbrayut** [comment]
Not sure... per the remarks at https://msdn.microsoft.com/en-us/library/windows/desktop/ms695279(v=vs.85).aspx I thought you still have to initialize once on each thread when using MTA.

**Comment by mattn** [comment]
However, if you fix issues with calling CoInitializeEx, please add it into your code. Maybe it may depend on that goroutine is not really thread.

**Comment by romanblachman** [comment]
This is still an open issue with Windows Server 2016 latest updates. Did anyone report this issue to Microsoft?

**Comment by mattn** [comment]
Did you try `runtime.LockOSThreads()` ?

**Comment by nook24** [comment]
I'm also affected by this :( Windows Server 2016
